W-13153401: Document "Iterative Scaffolding" (#47)
mulesoft/docs-code-builder: modules/ROOT/pages/implement-local-api-specification.adoc

= Implement a Local API Specification

> Anypoint Code Builder enables you to scaffold your API specification into a Mule project without having to publish it to Exchange. You can work on your Mule project then edit the API specification, and re-scaffold it to reflect your new changes.

Reviewer: ```suggestion
Anypoint Code Builder enables you to scaffold an API specification into a Mule project without having to publish it to Exchange. If you edit the API specification, re-scaffold it to reflect the changes in the Mule project.
```
Please confirm if this is accurate

> Both API design and API implementation tasks are now synchronized. You can then publish your API specification and deploy your Mule project.

Reviewer: Is the first sentence stating that this happened or just showing this as a feature of ACB? If the latter: 
```suggestion
Anypoint Code Builder synchronizes API design and API implementation tasks and enables you to publish API specifications and deploy Mule projects.
```

> Both API design and API implementation tasks are now synchronized. You can then publish your API specification and deploy your Mule project.

Reviewer: Another option for this could be: 
```suggestion
After you synchronize API design and API implementation tasks, publish the API specification and deploy the Mule project.
```

When you are designing an API specification, click the *Implement this local API* icon to scaffold your current API specification into a new Mule project:

image::implement-local-api-icon.png[]

> Code Builder creates a workspace with both your API specification and Mule project and keeps them synchronized.

Reviewer: ```suggestion
Code Builder creates a workspace with both the API specification and Mule project and keeps them synchronized.
```

[NOTE]

> Additionally, you can open the Command Palette (`ctrl/cmd + shift + P`) and type _MuleSoft: Implement this Local API_.

Reviewer: ```suggestion
Optionally, you can open the Command Palette (`ctrl/cmd + shift + P`) and type _MuleSoft: Implement this Local API_.
```
I think this is more accurate, please confirm

> With your API specification and related Mule project in sync, you can now modify your API specification and implement your changes simultaneously.

Reviewer: ```suggestion
After the API specification and the related Mule project are synced, you can modify the API specification and implement changes simultaneously.
```

> By iterating between your API design and implementation tasks, you can speed up your development process.

Reviewer: ```suggestion
By iterating between the API design and implementation tasks you can speed up the development process.
```

> == Working Iteratively Between Your API Specification and Mule Project

Reviewer: ```suggestion
== Working Iteratively Between an API Specification and a Mule Project
```

To implement a local API:

. Open an API specification. +
You can either import an API from Design Center or design a new API specification.
. Click the *Implement this local API* icon.

> . Code Builder creates a multi-root workspace to keep your API specification and Mule project in sync. +

Reviewer: ```suggestion
+
Code Builder creates a multi-root workspace to keep your API specification and Mule project in sync. +
```
Informational, not an actual step

A multi-root workspace enables you to work with both your API specification and Mule project folders at the same time.

> . After the workspace is created, Code Builder builds your API. +
> In this step, Code Builder packages the API specification and publishes it to a local Maven repository as a snapshot.

Reviewer: ```suggestion
+
After creating the workspace, Code Builder packages the API specification and publishes it to a local Maven repository as a snapshot.
```
It gets a bit confusing to differentiate the actual steps to take from the background processes that ACB runs.

. Choose a location to store your workspace. +

> Your multi-root workspace is a `.code-workspace` file extension that keeps both of your directories coordinates.

Reviewer: ```suggestion
The multi-root workspace is a file with the `.code-workspace` extension that keeps both of your directories coordinated.
```

+
[IMPORTANT]
--

> Save your `.code-workspace` file in your home directory. Do not save it inside your API specification folder.

Reviewer: ```suggestion
Save the `.code-workspace` file in your home directory. Do not save it inside the API specification folder.
```

--
. Choose a project name for your Mule project.
. Choose a location for your Mule project.

> . Code Builder scaffolds your current local API specification into a new Mule project.

Reviewer: ```suggestion
+
Code Builder then scaffolds your current local API specification into a new Mule project.
```

You can now start iterating between your API specification and its related Mule project.

> Every time you modify your API specification, you can run a *Re-Scaffold this local API* command to update your Mule project based on the changes in your Mule application. For example, if you add a new endpoint to your API specification, Code Builder scaffolds a new flow for that endpoint.

Reviewer: ```suggestion
Every time you modify an API specification, you can run a *Re-Scaffold this local API* command to update the  Mule project based on the changes in the Mule application. For example, if you add a new endpoint to an API specification, Code Builder scaffolds a new flow for that endpoint.
```

> You can close your multi-root workspace by running the *Workspaces: Close Workspace* command. You can re-open your synchronized project by opening the workspace you created when you started implementing your local API.

Reviewer: ```suggestion
To close your multi-root workspace, run the *Workspaces: Close Workspace* command. 

To re-open the synchronized project, open the workspace you created when you started implementing your local API.
```
